design-proposals/2022-09-02-design-proposal-template.md

```diff
@@ -1,21 +1,155 @@
+
+12. Developers should be able to associate specific styles with specific layer rendering.  The style sheet should be able to call out the type of representation it would like beyond the default.
+
+13. Developers should be able to add new visual representations fed by existing geometry types.  Data display driven by tile sets, for example.
+
+14. We should be able to drop out unused layer types with a compile flag.
 
 ## Proposed Change
 
-Describe what you would like to do to solve the problem. This part is the actual design proposal and can be extensive with multiple subsections.
+### Shader Representation
+
+We need an external representation for shaders so they can be added or replaced by developers.
+
+The shaders need a representation visible outside the toolkit and we need to change the way they’re handled inside the toolkit.  We need to be able to replace them, add brand new ones, and allow developers to control their shaders.
+
+#### The way it is now
+
+Individual shaders are represented by an object class with massive amounts of template logic and a minimum of in-line comments.  They’re opaque from the outside and only controllable through vector tile data and styles. 
+
+The [render_raster_layer](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/layers/render_raster_layer.cpp), as an example, asks for the instantiation of (eventually) RasterProgram.  Rather than having the source for the program, that then pokes into a compressed chunk of memory that contains the source, which is then uncompressed and fed into OpenGL for compilation.
+
+All of this makes sense in context, but is completely inaccessible outside the toolkit and rather incomprehensible when making changes.
+
+#### Required changes
+To let developers control their own shaders we’ll need to:
+- Allow shaders to be instantiated with shader source (one option).
+- Allow shaders to be instantiated by reference to unique names in pre-compiled file (Metal does this).
+- Set uniform values by name
+- Set uniform structs by name (Metal uses structs)
+
+#### Benefits
+We'll be able to:
+- Create entirely new shaders to do new and different things.
+- Replace existing shaders without having to recompile the whole set.
+- Let developers communicate with their own shaders.
+
+#### For GLES we’ll need to:
+- Push aside the existing Program hierarchy and rename it with a GLES extension.
+- Allow for named uniforms for new Programs
+- We can probably ignore that requirement for the existing shaders and just wrap them
+
+### Shader Registry
+
+Once developers can create their own shaders, we’ll need a way of managing them, thus the Shader Registry.  This registry will associate a given shader (Program) with a name.  Basic functionality will include:
+- Add a new shader by name
+- Replace an existing shader by name
+- Query a shader by name
+- Publish the well known names of the core shaders for the basic functionality
```

Comment:
Just a suggestion, writing a whole shader is a little bit harder for the end user, what about providing a shader interpolate interface, like mutating the shaders' color output, and mutating the line's width depending on line distance.

Reply:
I hope we can make writing a whole shader easier by exposing the shader source and wiring up the inputs dynamically.  So you could replace a whole shader more simply.

But this is an interesting idea.  I don't think we'd want to try it in this PR, but it deserves some thought.

Reply:
Maybe it could also help to make a few examples. Then people can take them and modify according to their needs...

Reply:
Yes!  That's a documentation idea.  Might make sense as part of the Metal effort.

Reply:
Great, maybe we can support such kind of feature in the future.